### PR TITLE
Query past contract events

### PR DESCRIPTION
**Pull request by emizzle**
Based on [ethers.js's `queryFilter`](https://github.com/ethers-io/ethers.js/blob/master/packages/contracts/src.ts/index.ts#L1039), allows querying of past contract events, by querying the logs for a contract's event topic.

### CODE REVIEW

File: testmodule/testContracts.nim [hunk 1]
```diff
@@ -188,3 +188,51 @@ for url in ["ws://localhost:8545", "http://localhost:8545"]:
...
+    test "can query last block event log":
+
+      let signer0 = provider.getSigner(accounts[0])
+
+      discard await token.connect(signer0).mint(accounts[0], 100.u256)
+      discard await token.connect(signer0).transfer(accounts[1], 50.u256)
+
+      let logs = await token.queryFilter(Transfer)
+
+      check eventually logs == @[
```
Comment (markspanbroek) [inline testmodule/testContracts.nim:201]: I don't think that you need `eventually` here?

```suggestion
      check logs == @[
```

File: testmodule/testContracts.nim [hunk 2]
```diff
@@ -188,3 +188,51 @@ for url in ["ws://localhost:8545", "http://localhost:8545"]:
...
+      discard await token.connect(signer0).mint(accounts[0], 100.u256)
+      discard await token.connect(signer0).transfer(accounts[1], 50.u256)
+      discard await token.connect(signer1).transfer(accounts[2], 25.u256)
+
+      let currentBlock = await provider.getBlockNumber()
+      let logs = await token.queryFilter(Transfer,
+                                         BlockTag.init(currentBlock - 3),
+                                         BlockTag.latest)
+
+      check eventually logs == @[
```
Comment (markspanbroek) [inline testmodule/testContracts.nim:219]:
```suggestion
      check logs == @[
```

File: testmodule/testContracts.nim [hunk 3]
```diff
@@ -188,3 +188,51 @@ for url in ["ws://localhost:8545", "http://localhost:8545"]:
...
+      let signer0 = provider.getSigner(accounts[0])
+      let signer1 = provider.getSigner(accounts[1])
+
+      discard await token.connect(signer0).mint(accounts[0], 100.u256)
+      discard await token.connect(signer0).transfer(accounts[1], 50.u256)
+      discard await token.connect(signer1).transfer(accounts[2], 25.u256)
+
+      let currentBlock = await provider.getBlockNumber()
+      let logs = await token.queryFilter(Transfer,
+                                         BlockTag.init(currentBlock - 3),
```
Comment (markspanbroek) [inline testmodule/testContracts.nim:216]: Wouldn't it be a more interesting test if you checked only 2 blocks back? To ensure that it takes into account the `from` block number?

**Review by markspanbroek — APPROVED** [review]
Very nice! Thanks @emizzle!

File: testmodule/testContracts.nim (hunk 3 above)
Comment (emizzle) [reply testmodule/testContracts.nim:216]: Yes, good idea. Updated.

**Pull request merged**